### ISSUE TITLE
tf.where raises ValueError for RaggedTensor arguments

### ISSUE THREAD

**Issue by Andreas5739738** [tensorflow/tensorflow#37351]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): no
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): OS X
- TensorFlow installed from (source or
binary): `pip install tensorflow`
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de410 2.1.0
- Python version: 3.7.4

**Describe the current behavior**
Calling tf.where with a RaggedTensor argument raises `ValueError: TypeError: object of type 'RaggedTensor' has no len()`.

**Describe the expected behavior**
No exception raised. According to https://www.tensorflow.org/api_docs/python/tf/ragged, tf.where is supposed to support RaggedTensor.

**Standalone code to reproduce the issue** 
```
import tensorflow as tf
digits = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
tf.where(tf.equal(digits, 1), 1, 0)
```

**Other info / logs** 
```
In [4]: tf.where(tf.equal(digits, 1), 1, 0)
---------------------------------------------------------------------------
_FallbackException                        Traceback (most recent call last)
/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_math_ops.py in select_v2(condition, t, e, name)
   8679         _ctx._context_handle, tld.device_name, "SelectV2", name,
-> 8680         tld.op_callbacks, condition, t, e)
   8681       return _result

_FallbackException: This function does not handle the case of the path where all inputs are not already EagerTensors.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-a793e17a60b8> in <module>
----> 1 tf.where(tf.equal(digits, 1), 1, 0)

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/array_ops.py in where_v2(condition, x, y, name)
   3928       return gen_array_ops.where(condition=condition, name=name)
   3929   elif x is not None and y is not None:
-> 3930     return gen_math_ops.select_v2(condition=condition, t=x, e=y, name=name)
   3931   else:
   3932     raise ValueError("x and y must both be non-None or both be None.")

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_math_ops.py in select_v2(condition, t, e, name)
   8683       try:
   8684         return select_v2_eager_fallback(
-> 8685             condition, t, e, name=name, ctx=_ctx)
   8686       except _core._SymbolicException:
   8687         pass  # Add nodes to the TensorFlow graph.

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_math_ops.py in select_v2_eager_fallback(condition, t, e, name, ctx)
   8706   _attr_T, _inputs_T = _execute.args_to_matching_eager([t, e], ctx)
   8707   (t, e) = _inputs_T
-> 8708   condition = _ops.convert_to_tensor(condition, _dtypes.bool)
   8709   _inputs_flat = [condition, t, e]
   8710   _attrs = ("T", _attr_T)

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py in convert_to_tensor(value, dtype, name, as_ref, preferred_dtype, dtype_hint, ctx, accepted_result_types)
   1312
   1313     if ret is None:
-> 1314       ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
   1315
   1316     if ret is NotImplemented:

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py in _constant_tensor_conversion_function(v, dtype, name, as_ref)
    315                                          as_ref=False):
    316   _ = as_ref
--> 317   return constant(v, dtype=dtype, name=name)
    318
    319

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py in constant(value, dtype, shape, name)
    256   """
    257   return _constant_impl(value, dtype, shape, name, verify_shape=False,
--> 258                         allow_broadcast=True)
    259
    260

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py in _constant_impl(value, dtype, shape, name, verify_shape, allow_broadcast)
    264   ctx = context.context()
    265   if ctx.executing_eagerly():
--> 266     t = convert_to_eager_tensor(value, ctx, dtype)
    267     if shape is None:
    268       return t

/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py in convert_to_eager_tensor(value, ctx, dtype)
     94       dtype = dtypes.as_dtype(dtype).as_datatype_enum
     95   ctx.ensure_initialized()
---> 96   return ops.EagerTensor(value, ctx.device_name, dtype)
     97
     98

ValueError: TypeError: object of type 'RaggedTensor' has no len()
```

**Comment by gadagashwini** [comment]
@Andreas5739738, I tried replicating the issue but getting different error. Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/4138a70a375e358fe5022fe3358db691/untitled424.ipynb).
Here 
```
tf.where(
    condition, x=None, y=None, name=None
)
```  x and y must be broadcastable to the same shape. Can you confirm the issue. Thanks!

**Comment by Andreas5739738** [comment]
It looks like the notebook is testing with TF 1.15 (I see `/tensorflow-1.15.0/python3.6/tensorflow_core/python/framework/tensor_util.py` in the error messages)?

Here is a different repro case where the tensor dimensions are the same, which results in the same error:
```
>>> tf.where(tf.equal(digits, 1),tf.ones_like(digits),tf.zeros_like(digits))
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_math_ops.py", line 8680, in select_v2
    tld.op_callbacks, condition, t, e)
tensorflow.python.eager.core._FallbackException: This function does not handle the case of the path where all inputs are not already EagerTensors.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/array_ops.py", line 3930, in where_v2
    return gen_math_ops.select_v2(condition=condition, t=x, e=y, name=name)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_math_ops.py", line 8685, in select_v2
    condition, t, e, name=name, ctx=_ctx)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_math_ops.py", line 8706, in select_v2_eager_fallback
    _attr_T, _inputs_T = _execute.args_to_matching_eager([t, e], ctx)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/eager/execute.py", line 263, in args_to_matching_eager
    t, dtype, preferred_dtype=default_dtype, ctx=ctx))
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py", line 1314, in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py", line 317, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py", line 258, in constant
    allow_broadcast=True)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py", line 266, in _constant_impl
    t = convert_to_eager_tensor(value, ctx, dtype)
  File "/usr/local/lib/python3.7/site-packages/tensorflow_core/python/framework/constant_op.py", line 96, in convert_to_eager_tensor
    return ops.EagerTensor(value, ctx.device_name, dtype)
ValueError: TypeError: object of type 'RaggedTensor' has no len()
```

**Comment by gadagashwini** [comment]
I could replicate the issue with tf 2.1.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/c29154968a86ae94e4973533b1385df1/untitled431.ipynb). Thanks

**Comment by edloper** [comment]
The semantics of tf.where changed from TF1 to TF2 (in particular the broadcasting semantics changed); and it looks like the ragged version of tf.where only overrides where_v1, not where_v2.  We will need to write an updated version of ragged_where with the v2 semantics.

As a stop-gap solution until then, you can use tf.compat.v1.where(), which does support ragged tensors.

**Comment by edloper** [comment]
Note: since v1 where doesn't do broadcasting, you may need to do it yourself, depending on what your use case looks like.  E.g., for the gist above we'd need  to use something like:

```
digits = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
tf.compat.v1.where(tf.equal(digits, 1), tf.ones_like(digits), tf.zeros_like(digits))
```

**Comment by ravinderkhatri** [comment]
I tried using below example in tf 2.2.0 and giving me the same error 
```
b = tf.ragged.constant([[True, False, True], [False, True]])
x = tf.ragged.constant([["A", "B", "C"], ["D", "E"]])
y = tf.ragged.constant([['a', "b", "c"], ["d", "e"]])

```
**ValueError: TypeError: object of type 'RaggedTensor' has no len()**

**Comment by edloper** [comment]
You tried using `tf.compat.v1.where` in tf 2.2.0?  The following works for me:

```
b = tf.ragged.constant([[True, False, True], [False, True]])
x = tf.ragged.constant([["A", "B", "C"], ["D", "E"]])
y = tf.ragged.constant([['a', "b", "c"], ["d", "e"]])
tf.compat.v1.where(b, x, y)
```

Note that I'm calling `tf.compat.v1.where`, and *not* `tf.where`.  As I said, it looks like the v2 version of `tf.where` hasn't been updated to handle ragged tensors yet, but we will try to update it soon.

**Comment by ravikyram** [comment]
I have tried in colab with TF versions 2.2,2.3-rc1,nightly versions(`2.4.0-dev20200712`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/22aa2a664bae02e3ae23f16e566fc4c7/untitled116.ipynb).Thanks!

**Comment by edloper** [comment]
Fixed in d753f3aff1ee4d23572e059f5a2d49269bbebac9.
